src/pcp.rs

```diff
@@ -135,65 +199,269 @@ pub trait ValueParam: Clone + Debug {
+                proof_len += next_len;
+
+                Ok(Box::new(QueryShimGadget::new(gadget, r, proof_data)?)
+                    as Box<dyn Gadget<Self::Field>>)
+            })
+            .collect::<Result<Vec<_>, _>>()?;
+
+        if proof_len < proof.len() {
+            return Err(PcpError::Query(format!(
+                "long proof (expected length {})",
+                self.proof_len(),
+            )));
+        }
+
+        if query_rand.len() > shim.len() {
+            return Err(PcpError::Query(format!(
+                "long query randomness: got {}; want {}",
+                query_rand.len(),
+                self.query_rand_len()
+            )));
+        }
+
+        // Create a buffer for the verifier data. This includes the output of the validity circuit and,
+        // for each gadget `shim[idx].inner`, the wire polynomials evaluated at the query randomness
+        // `query_rand[idx]` and the gadget polynomial evaluated at `query_rand[idx]`.
+        let data_len = 1
+            + (0..shim.len())
+                .map(|idx| shim[idx].arity() + 1)
+                .sum::<usize>();
+        let mut verifier = Vec::with_capacity(data_len);
+
+        // Run the validity circuit with a sequence of "shim" gadgets that record the inputs to each
+        // wire for each gadget call. Record the output of the circuit and append it to the verifier
+        // message.
+        //
+        // NOTE The proof of [BBC+19, Theorem 4.3] assumes that the output of the validity circuit is
+        // equal to the output of the last gadget evaluation. Here we relax this assumption. This
+        // should be OK, since it's possible to transform any circuit into one for which this is true.
+        // (Needs security analysis.)
+        let validity = self.valid(&mut shim, input, joint_rand, num_shares)?;
+        verifier.push(validity);
+
+        // Fill the buffer with the verifier message.
+        for idx in 0..shim.len() {
+            let r = query_rand[idx];
+            let gadget = shim[idx]
+                .as_any()
+                .downcast_ref::<QueryShimGadget<Self::Field>>()
+                .unwrap();
+
+            // Reconstruct the wire polynomials `f[0], ..., f[g_arity-1]` and evaluate each wire
+            // polynomial at query randomness `r`.
+            let m = (1 + gadget.calls()).next_power_of_two();
+            let m_inv =
+                Self::Field::from(<Self::Field as FieldElement>::Integer::try_from(m).unwrap())
+                    .inv();
+            let mut f = vec![Self::Field::zero(); m];
+            for wire in 0..gadget.arity() {
+                discrete_fourier_transform(&mut f, &gadget.f_vals[wire], m)?;
+                discrete_fourier_transform_inv_finish(&mut f, m, m_inv);
+                verifier.push(poly_eval(&f, r));
+            }
+
+            // Add the value of the gadget polynomial evaluated at `r`.
+            verifier.push(gadget.p_at_r);
+        }
```

Comment:
In the unlikely event that `poly_eval` returns a value longer than `shim[idx].arity`, we will end up with more than `data_len` elements in `verifier`. Could we check that `verifier.len() == self.verifier_len()` and error out if so?

Reply:
Hmm, I'm not sure I understand the problem. Are you concerned about an overflow? `poly_eval` always returns exactly one field element. In any case, if a `.push()` would exceed the limits of the memory allocated for the vector, the vector's memory will be extended.

Reply:
Oh, good point. My thinking is that this module should defend itself against errors in implementation of a `Type`, but you're right, `poly_eval` can't be implemented wrong, so we have a reasonable guarantee that the vector will contain the right number of elements. It might still be nice to check that the `Type`'s `verifier_len()` implementation is correct and panic -- your call.

Reply:
I'll add an assertion, like we do for proof length in `prove`.